doc/ADLSgen2 implementation guidance.md

```diff
@@ -0,0 +1,134 @@
+
+### ADLSgen2 in Azurite
+#### Current status: 
+<table>
+<tr>  
+<th></th>
+<th>FNS account</th>
+<th>HNS account</th>
+</tr>
+<tr>
+<td>Blob Endpoint API</td>
+<td>Already implemented in Azurite.
+  
+(The customer PR1 is mostly in refine and revise blob API implementation.)</td>
+<td>Not in Azurite, Phase II in below plan.
+  
+Behavior similar as FNS blob, but should have a little different on API, and performance different</td>
+</tr>
+<tr>
+<td>Dfs Endpoint API</td>
+<td>Not in Azurite, Phase I in below plan.
+  
+Support most DFS API, but some action not supported and API behavior different, and performance/atomic different</td>
+<td>Not in Azurite, Phase II in below plan
+
+Support all DFS APIs, including ACL/permission support.
+
+(The customer PR2 is mostly on these APIs.)</td>
+</tr>
+</table>
+
+#### Implementation Plan:
+##### Phase I: implementation DFS API on FNS account 
+1. This should be much simpler than HNS account implementation.
+    1. Current Azurite is based on FNS account, so: 
+        1. Don’t need to change data store structure.
+        2. Don’t need Azurite user to differ HNS/FNS account.
+           
+2. The change will add all dfs API interface to Azurite, which can help to support phase II.
+   
+4. Code change should be split into several small PRs as following:
+    1. 1 PR to add dfs swagger and the auto-generated API interface (no manual change on auto generated code)
+    2. 1 PR to add DFS endpoint 
+    3. Several PRs to implement each dfs API (with credential handler), include testing
+       
+5. Need make sure each API behavior is aligned on rest API doc , also aligned with real Azure Server. See more in validation criteria.
+   
+7. The blob/dfs endpoint should share same data store, talk to same instance of BlobLokiMetadataStore & BlobSqlMetadataStore
+   
+9. Need to work with Azure Storage SDKs to change and support new dfs port (say: 10004)
+    1. E.g. .Net SDK need changes the blob/dfs Uri convert function in [this file](https://github.com/Azure/azure-sdk-for-net/blob/e8c40cc204b8cf750fcc820eab90d11f80612c3a/sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeUriBuilder.cs#L275)
+       
+##### Phase II: implementation Blob/DFS on HNS account
+1. Azurite user need configure each Azurite Account type as HNS/FNS when Azurite starts up. 
+    1. Need design how to input the config (default should be FNS)
+    2.How to handle it when user start Azurite with change account type? (Report error? )
```

Comment:
I'll suggest Azurite doesn't support FNS/HNS migration and account type change. The storage account type is finalized after creation. A wrong configuration (like wrong HNS/FNS type) should get error reported.


```suggestion
    2. How to handle it when user start Azurite with change account type? (Report error? )
```

Reply:
Add to Phase II 1.iii.